### DISCUSSION TITLE
Dependency constraints for rule tests

### DISCUSSION THREAD

**Discussion by bradzacher** [eslint/eslint#16395] [Ideas]
I wanted to float this idea past everyone here. Happy to raise an RFC for this if there's appetite for the change.

Recently I was working on adding regression testing to the @typescript-eslint project. We have a few dependencies for which we support both current and old versions (for us specifically typescript and eslint).

I was [working on adding some CI jobs to test against old versions](https://github.com/typescript-eslint/typescript-eslint/pull/5573) when I realised that many of our rule tests need to be run against specific versions of dependencies.

Namely we have two cases:
- some use newer TS syntax, and thus will crash on old versions of TS
- we maintain some "extension rules" which are built directly on top of ESLint rules - meaning their exact behaviour depends on the ESLint version being used

[In some other repos I've seen them handle this dependency problem by declaring separate test blocks](https://github.com/import-js/eslint-plugin-import/blob/753505d062f7ae6736eb4ffb2cc35c3f699c5b6f/tests/src/rules/consistent-type-specifier-style.js#L358-L360) for example:

```ts
const ruleTester = new RuleTester();

// tests that are good for any version
ruleTester.run('rule', rule, { ... });

if (semver.satisfies(require('my-dep/package.json').version, '>=7')) {
  // tests that require specific versions
  ruleTester.run('rule', rule, { ... });
}
```

But I don't really like this style because it means you're splitting tests up across multiple calls.

Instead I [built out some infrastructure](https://github.com/typescript-eslint/typescript-eslint/pull/5750) directly into the rule tester that allows you to constrain specific rule tests to specific versions within the test's declaration:
```ts
ruleTester.run('my-rule', rule, {
  valid: [
    {
      code: 'const x = 1;',
      dependencyConstraints: {
        // enforces the semver ">=7.32.0"
        eslint: '7.32.0',

        // object allows the user to fully control the semver.satisfies call
        'some-other-module': {
          type: 'semver',
          version: '~6',
          options: {
            includePrerelease: false,
          },
        },
      },
    },
  ],
  invalid: [],
});
```

Do we think this might be more generally useful to the plugin author community?
Is this something that might be worth building directly into ESLint?
As I said - happy to raise an RFC for this if there's appetite for the change.

**Comment by nzakas** [comment]
Sorry for the delay on responding to this. In general, I see `RuleTester` as a lower-level primitive rather than a kitchen sink solution for rule testing. So, I'd like to avoid adding too much more functionality, especially if it ends up being Node.js-specific as we are looking at moving towards being [runtime agnostic](https://github.com/eslint/eslint/issues/16482).

That said, it may be useful to share with others.
